# VarCall2015
Materials for the VarCall2015 course

##NGS for evolutionary biologists: from basic scripting to variant calling
CINECA Auditorium, via dei Tizii 2/C, Rome - 23-27 November 2015

Recent advances in genome sequencing technologies provide unprecedented opportunities to characterise individual genomes and identify mutations relevant for the study of both evolutionary questions and diagnosis and therapy. This course runs over five days focusing on variant analysis of NGS data and is addressed to wet lab scientists with no previous knowledge of the use of a command line, statistical computing packages, and programming. Participants are immersed in the experience of running a complete NGS data analysis workflow, from the retrieval of raw data from bioinformatics resources, to the downstream analysis of the processed data.

This repository is for the course materials and it is organised as follows:

- Day 1: [Unix/Linux command line](./1-unix_command_line) 
- Day 1: [NGS and Population Genomics](./2-ngs_and_population_genomics)
- Day 2: [NGS platforms_and_dataQC](./3-ngs_platforms_and_dataQC)
- Day 3: [Mapping and BAM refinement](./4-mapping_and_bam_refinment)
- Day 4: [Variant Calling](./5-variant_calling)
- Day 5: [Projects](./6-projects)

## Projects
- [Before we start](Projects/00-beforewestart_sum.md)
- [Project 1](Projects/01-Project-01_sum.md)
- [Project 2](Projects/02-Project-02_sum.md)

## Tutorials
- [T0: before we start](Tutorials/T0_before_we_start.md)
- [T3: Data QC](Tutorials/T3_dataQC.md)
- [T4: Mapping `.bam` refinement](Tutorials/T4_mapping_and_bam_refinement.md)
- [T5.1: Variant Calling (SNPs)](Tutorials/T5.1_variantcalling_snps_tutorial.md)
- [T5.2: Variant Calling (SVs)](Tutorials/T5.2_variantcalling_stucturalvariants_tutorial.md)
- [T5.3: Variant Calling and Filtering - Exercise](Tutorials/T5.3_variantcalling_filtering_exercises.md)
